diff --git a/Cargo.toml b/Cargo.toml
index 2ee0ba8..1b81709 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "plxsversion"
-version = "2.2.0"
+version = "2.3.0"
 edition = "2024"
 
 [dependencies]
diff --git a/plxsversion.cmake b/plxsversion.cmake
index 6b88f06..4c0d395 100644
--- a/plxsversion.cmake
+++ b/plxsversion.cmake
@@ -45,7 +45,7 @@ function(plxsversion_create_target)
   cmake_parse_arguments(
     VER
     "PRINT;TIME"
-    "LANG;SOURCE;INPUT;TARGET_SUFFIX"
+    "LANG;SOURCE;INPUT;TARGET_SUFFIX;NAMESPACE;INCLUDE_PREFIX"
     ""
     ${ARGN}
   )
@@ -59,6 +59,17 @@ function(plxsversion_create_target)
     list(APPEND OPTIONS "--time")
   endif()
 
+  if(VER_NAMESPACE)
+    list(APPEND OPTIONS "--namespace" ${VER_NAMESPACE})
+    set(VERSION_LIBRARY "plxsversion-${VER_NAMESPACE}")
+  else()
+    set(VERSION_LIBRARY plxsversion)
+  endif()
+
+  if(VER_INCLUDE_PREFIX)
+    list(APPEND OPTIONS "--include-prefix" ${VER_INCLUDE_PREFIX})
+  endif()
+
   if(NOT VER_LANG)
     set(VER_LANG cpp)
   endif()
@@ -79,23 +90,41 @@ function(plxsversion_create_target)
     set(VER_INPUT ${CMAKE_CURRENT_SOURCE_DIR})
   endif()
 
-  if(NOT VER_TARGET_SUFFIX)
-    set(VERSION_LIBRARY plxsversion)
+  # if(NOT VER_TARGET_SUFFIX)
+  #   set(VERSION_LIBRARY plxsversion)
+  # else()
+  #   set(VERSION_LIBRARY "plxsversion-${VER_TARGET_SUFFIX}")
+  # endif()
+
+  if(VER_INCLUDE_PREFIX)
+    set(REL_OUT_PATH "version.hpp")
+    if(${VER_LANG} STREQUAL "c")
+      set(REL_OUT_PATH "version.h")
+    endif()
+    set(INCLUDE_DIR ${CMAKE_CURRENT_BINARY_DIR})
   else()
-    set(VERSION_LIBRARY "plxsversion-${VER_TARGET_SUFFIX}")
+    _set_relative_out_file_path(${VER_LANG})
+    set(INCLUDE_DIR ${CMAKE_CURRENT_BINARY_DIR}/plxs)
   endif()
 
-  _set_relative_out_file_path(${VER_LANG})
   set(OUT_FILE "${CMAKE_CURRENT_BINARY_DIR}/${REL_OUT_PATH}")
   _create_version_file(${VER_LANG} ${VER_SOURCE} ${VER_INPUT} ${OUT_FILE} ADDITIONAL_OPTIONS ${OPTIONS})
 
   add_library(${VERSION_LIBRARY} INTERFACE)
-  target_include_directories(${VERSION_LIBRARY} 
-    INTERFACE 
-      $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/plxs>
-      $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/plxs>)
+  target_include_directories(${VERSION_LIBRARY}
+    INTERFACE
+      $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+      $<BUILD_INTERFACE:${INCLUDE_DIR}>)
   message(STATUS "${VERSION_LIBRARY} created.")
 
   set_property(TARGET ${VERSION_LIBRARY} APPEND PROPERTY ADDITIONAL_CLEAN_FILES "${OUT_FILE}")
-  _set_version_cmake_variable(PLXSVERSION_STRING ${OUT_FILE})
+  if(VER_INCLUDE_PREFIX)
+    set(VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/${VER_INCLUDE_PREFIX}/version.hpp")
+    if(${VER_LANG} STREQUAL "c")
+      set(VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/${VER_INCLUDE_PREFIX}/version.h")
+    endif()
+    _set_version_cmake_variable(PLXSVERSION_STRING ${VERSION_FILE})
+  else()
+    _set_version_cmake_variable(PLXSVERSION_STRING ${OUT_FILE})
+  endif()
 endfunction(plxsversion_create_target)
\ No newline at end of file
diff --git a/src/version_builder/__main__.py b/src/version_builder/__main__.py
index f8baca4..9aa0b6b 100644
--- a/src/version_builder/__main__.py
+++ b/src/version_builder/__main__.py
@@ -32,6 +32,19 @@ def execute() -> None:
         action=argparse.BooleanOptionalAction,
         help="include time data in the version infomation",
     )
+    parser.add_argument(
+        "--namespace",
+        "-n",
+        required=False,
+        default="plxsversion",
+        help="C++ namespace to put the version info into",
+    )
+    parser.add_argument(
+        "--include-prefix",
+        required=False,
+        default="",
+        help="subdirectory to put the generated file into",
+    )
     parser.add_argument(
         "--cargo",
         "-c",
@@ -52,6 +65,8 @@ def execute() -> None:
         source_input=args.input,
         output_file=args.file,
         lang=args.lang,
+        namespace=args.namespace,
+        include_prefix=args.include_prefix,
         optional_config=main.OptionalConfiguration(
             print_created_file=args.print, include_time=args.time, cargo_version=args.cargo
         ),
diff --git a/src/version_builder/formatter.py b/src/version_builder/formatter.py
index 460aeca..d4313b7 100644
--- a/src/version_builder/formatter.py
+++ b/src/version_builder/formatter.py
@@ -1,12 +1,12 @@
 from version_builder.version_data import VersionData
 
 
-def to_cpp(version_data: VersionData) -> str:
-    return _CppFormatter().format(version_data)
+def to_cpp(version_data: VersionData, *, namespace: str, include_prefix: str) -> str:
+    return _CppFormatter(namespace=namespace, include_prefix=include_prefix).format(version_data)
 
 
-def to_cpp11(version_data: VersionData) -> str:
-    return _Cpp11Formatter().format(version_data)
+def to_cpp11(version_data: VersionData, *, namespace: str, include_prefix: str) -> str:
+    return _Cpp11Formatter(namespace=namespace, include_prefix=include_prefix).format(version_data)
 
 
 def to_c(version_data: VersionData) -> str:
@@ -25,10 +25,20 @@ class _Formatter:
         return self.main_formatter(version_data)
 
 
+def _get_include_guard(include_prefix: str, namespace: str) -> str:
+    guard = f"{include_prefix.replace('/', '_')}_{namespace.replace('::', '_')}_VERSION_HPP"
+    return guard.upper()
+
+
 # ----------------------------------------
 # C++ Formatter
 # ----------------------------------------
 class _CppFormatter(_Formatter):
+    def __init__(self, *, namespace: str, include_prefix: str) -> None:
+        super().__init__()
+        self.namespace = namespace
+        self.include_guard = _get_include_guard(include_prefix, namespace)
+
     def main_formatter(self, version_data: VersionData) -> str:
         return f"""
 // ---------------------------------------------------
@@ -36,13 +46,13 @@ class _CppFormatter(_Formatter):
 // DO NOT MODIFY!
 // ---------------------------------------------------
 
-#ifndef PLXSVERSION_VERSION_HPP
-#define PLXSVERSION_VERSION_HPP
+#ifndef {self.include_guard}
+#define {self.include_guard}
 
 #include <cstdint>
 #include <string_view>
 
-namespace plxsversion {{
+namespace {self.namespace} {{
 
 inline constexpr std::string_view VERSION {{ "{version_data.qualified_version:s}" }};
 inline constexpr unsigned int MAJOR {{ {version_data.major:d} }};
@@ -57,9 +67,9 @@ inline constexpr bool DIRTY_BUILD {{ {str(version_data.is_dirty).lower():s} }};
 inline constexpr bool DEVELOPMENT_BUILD {{ {str(version_data.is_development_build).lower():s} }};
 inline constexpr std::string_view BUILD_METADATA {{ "{version_data.full_build_metadata:s}" }};
 {self._optional_output(version_data):s}
-}} // namespace plxsversion
+}} // namespace {self.namespace}
 
-#endif // PLXSVERSION_VERSION_HPP
+#endif // {self.include_guard}
 """
 
     def _optional_output(self, version_data: VersionData) -> str:
@@ -73,6 +83,13 @@ inline constexpr std::string_view BUILD_METADATA {{ "{version_data.full_build_me
 # C++11 Formatter
 # ----------------------------------------
 class _Cpp11Formatter(_Formatter):
+    def __init__(self, *, namespace: str, include_prefix: str) -> None:
+        super().__init__()
+        self.namespace = namespace
+        self.include_guard = _get_include_guard(include_prefix, namespace)
+        self.open_namespace = "namespace " + " { namespace ".join(self.namespace.split("::")) + " {"
+        self.close_namespace = "} " * len(self.namespace.split("::")) + f"// namespace {self.namespace}"
+
     def main_formatter(self, version_data: VersionData) -> str:
         return f"""
 // ---------------------------------------------------
@@ -80,12 +97,12 @@ class _Cpp11Formatter(_Formatter):
 // DO NOT MODIFY!
 // ---------------------------------------------------
 
-#ifndef PLXSVERSION_VERSION_HPP
-#define PLXSVERSION_VERSION_HPP
+#ifndef {self.include_guard}
+#define {self.include_guard}
 
 #include <cstdint>
 
-namespace plxsversion {{
+{self.open_namespace}
 
 constexpr const char *VERSION {{ "{version_data.qualified_version:s}" }};
 constexpr unsigned int MAJOR {{ {version_data.major:d} }};
@@ -100,9 +117,9 @@ constexpr bool DIRTY_BUILD {{ {str(version_data.is_dirty).lower():s} }};
 constexpr bool DEVELOPMENT_BUILD {{ {str(version_data.is_development_build).lower():s} }};
 constexpr const char *BUILD_METADATA {{ "{version_data.full_build_metadata:s}" }};
 {self._optional_output(version_data):s}
-}} // namespace plxsversion
+{self.close_namespace}
 
-#endif // PLXSVERSION_VERSION_HPP
+#endif // {self.include_guard}
 """
 
     def _optional_output(self, version_data: VersionData) -> str:
@@ -153,7 +170,7 @@ static const char *BUILD_METADATA = "{version_data.full_build_metadata:s}";
 #endif // PLXSVERSION_VERSION_H
 """
 
-    def _optional_output(self, version_data: VersionData) -> str:
+    def _optional_output(self, version_data: VersionData) -> str:.
         optional_output = ""
         if version_data.time:
             optional_output += f"""static const char *UTC_TIME = "{version_data.time:s}";\n"""
diff --git a/src/version_builder/main.py b/src/version_builder/main.py
index 6dad23b..40db4f3 100644
--- a/src/version_builder/main.py
+++ b/src/version_builder/main.py
@@ -13,7 +13,14 @@ class OptionalConfiguration:
 
 
 def create_version_file(
-    source: str, source_input: str, output_file: str, lang: str, *, optional_config: OptionalConfiguration = None
+    source: str,
+    source_input: str,
+    output_file: str,
+    lang: str,
+    *,
+    namespace: str,
+    include_prefix: str,
+    optional_config: OptionalConfiguration = None,
 ) -> None:
     if optional_config is None:
         optional_config = OptionalConfiguration()
@@ -29,6 +36,8 @@ def create_version_file(
         version_info=version_info,
         output_file=PosixPath(output_file),
         lang=lang,
+        namespace=namespace,
+        include_prefix=include_prefix,
         print_created_file=optional_config.print_created_file,
     )
 
@@ -46,15 +55,26 @@ def _get_version(source: str, source_input: str) -> version_data.VersionData:
 
 
 def _output_version_file(
-    version_info: version_data.VersionData, output_file: str, lang: str, *, print_created_file: bool
+    version_info: version_data.VersionData,
+    output_file: str,
+    lang: str,
+    *,
+    namespace: str,
+    include_prefix: str,
+    print_created_file: bool,
 ) -> None:
     """Convert version info into a requested format and outputs to a file."""
+    if include_prefix:
+        output_path = PosixPath(output_file).parent / include_prefix
+        output_path.mkdir(parents=True, exist_ok=True)
+        output_file = output_path / PosixPath(output_file).name
+
     match lang:
         case "cpp":
-            output = formatter.to_cpp(version_info)
+            output = formatter.to_cpp(version_info, namespace=namespace, include_prefix=include_prefix)
             expected_file_extension = ".hpp"
         case "cpp11":
-            output = formatter.to_cpp11(version_info)
+            output = formatter.to_cpp11(version_info, namespace=namespace, include_prefix=include_prefix)
             expected_file_extension = ".hpp"
         case "c":
             output = formatter.to_c(version_info)
